proposals/CodifiedVUs.adoc

```diff
@@ -0,0 +1,170 @@
+constructs, and back references such as "that element" or "its".
+The result is hard to interpret and sometimes ambiguous VUs.
+
+In essence, these VUs are basic algorithms composed of loops and conditions
+leading to assertions.
+However, a natural language (such as English) is not best suited to describe this.
+The authors of the spec, its users and the implementers of the Vulkan
+Validation Layers (VVL) are typically all programmers, which makes use of prose
+as the medium of communicating assertions even more awkward.
+
+== Proposal
+
+We could do better by writing VUs in a programming language instead.
+In the above example, one can identify a for loop, two conditions and an
+assertion.
+That could be written as:
+
+[source,python]
+~~~~
+  * [[VUID-VkRenderingInfo-colorAttachmentCount-06097]]
+    for attachment in pColorAttachments:
+     if (attachment.imageView != VK_NULL_HANDLE and
+         attachment.resolveMode != VK_RESOLVE_MODE_NONE):
+      resolveLayout = attachment.resolveImageLayout
+      require(resolveLayout != VK_IMAGE_LAYOUT_DEPTH_READ_ONLY_STENCIL_ATTACHMENT_OPTIMAL and
+              resolveLayout != VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_STENCIL_READ_ONLY_OPTIMAL)
+~~~~
+
+The benefits of writing the VU in this way are:
+
+  * Clear succinct logic:
+  ** The VUs are unambiguous, and easier to both write and read by programmers
+  ** Using named references (i.e. variables) obviates the need for
+     hard-to-follow back-references through pronouns.
+  ** "Built-in" functionality can be defined once in an appendix, and
+     referenced by VUs
+  * The VU can be type-checked: This prevents mistakes when writing the VU
+  * Code generation for VVL becomes a possibility
+
+Python is chosen as a language that is very easy to write and read, has minimal
+brace clutter and reduces vertical space by using indentation to create scope.
+Additionally, Python's `ast` module is capable of parsing Python code and
+greatly facilitates the job of tree traversal and code generation.
+
+A subset of the language is sufficient to describe the VUs.
+The abstract grammar of Python can be found in
+https://docs.python.org/3/library/ast.html[the documentation of the `ast` module].
+Of the defined nodes, the following should be enough for the purposes of Vulkan:
+
+  * `Module`, `Expression`
+  * `For`, `If`, `While`, `Break`, `Continue`
+  * Operators:
+  ** `BoolOp`: `And`, `Or`
+  ** `UnaryOp`: `Invert`, `Not`, `UAdd`, `USub`
+  ** `BinOp`: `Add`, `Sub`, `Mult`, `Div`, `Mod`, `Pow`, `LShift`, `RShift`,
+             `BitOr`, `BitXor`, `BitAnd`, `FloorDiv`
+  ** `IfExp`
+  ** `Compare`: `Eq`, `NotEq`, `Lt`, `LtE`, `Gt`, `GtE`
+  * `Attribute`
+  * `Call`
+  * `Assign`
+  * `Name`
+
+Additionally, a number of predefined functions (built-ins) are necessary to be
+able to describe the VUs.
+Examples include:
+
+- `require`: The equivalent of `assert`, describes the condition that Vulkan
+  requires.
```

Comment:
What is the advantage of having `require()` over the VU simply returning true or false?

Reply:
Are you suggesting turning `require(cond)` to `if not cond: return false`? `require()` is simpler IMO.

Reply:
WTF, of course not. Proper way to write `if not cond: return false` would be `return cond`.

Reply:
Except that doesn't allow you to support multiple `require` calls in the VU if needed.

See this very simple example: https://github.com/KhronosGroup/Vulkan-Docs/pull/2064/commits/af37bc96975071abde37d3e43d2dda6558f7ebbd#diff-09069e339671f4a6dd446abad39770bcb7b8907bcbb780ed636ec822e38da95d

> if we inplicitly assume every VU is boolean expression

That is one idea, but it'll be too restrictive compared with allowing the VU to have any logic.

Reply:
Maybe that shouldn't be a thing. If more than one is required, it points to some problem with it (e.g. the VU not being atomic enough).

Reply:
Re PS: That VVL could tell you _which_ of those requires failed, which makes it easier for the reader to understand where exactly the problem is. If VVL says `a && b` fails, you can't immediately see if it's `a` or `b` that has failed.

Re PPS: Yes, indeed that's what I have in mind. It can make the VUs a bit less repetitive.